metamorph/src/main/java/org/metafacture/metamorph/Metamorph.java

```diff
@@ -306,28 +310,37 @@ public void closeStream() {
         outputStreamReceiver.closeStream();
     }
 
-    protected void dispatch(final String path, final String value, final List<NamedValueReceiver> fallback) {
-        final List<NamedValueReceiver> matchingData = findMatchingData(path, fallback);
-        if (null != matchingData) {
-            send(path, value, matchingData);
-        }
-    }
-
-    private List<NamedValueReceiver> findMatchingData(final String path, final List<NamedValueReceiver> fallback) {
-        final List<NamedValueReceiver> matchingData = dataRegistry.get(path);
+    protected void dispatch(final String path, final String value, final List<NamedValueReceiver> fallbackReceiver) {
+        List<NamedValueReceiver> matchingData = dataRegistry.get(path);
+        boolean fallback = false;
         if (matchingData == null || matchingData.isEmpty()) {
-            return fallback;
+            fallback = true;
+            matchingData = fallbackReceiver;
+        }
+        if (null != matchingData) {
+            send(path, value, matchingData, fallback);
         }
-        return matchingData;
     }
 
-    private void send(final String key, final String value, final List<NamedValueReceiver> dataList) {
+    private void send(final String path, final String value, final List<NamedValueReceiver> dataList, final boolean fallback) {
         for (final NamedValueReceiver data : dataList) {
+            String key=path;
+            if (fallback && value != null && passEntityEvents) {
+                if (flattener.getCurrentEntityName() != null) {
+                    outputStreamReceiver.startEntity(flattener.getCurrentEntityName());
+                    key = literalPatternOfEntityMarker.split(path)[1];
```

Comment:
Previously (before 758327d) this would fail with an `ArrayIndexOutOfBoundsException` since the pattern didn't match a top-level literal path. Is there still a scenario where an entity name is present, but the path doesn't contain the entity marker?

Reply:
Not sure if I understand you. The check is necessary and your (good) proposal to [test top-level literal behaviour](https://github.com/metafacture/metafacture-core/pull/333#discussion_r502282475) would otherwise fail. And the check also prevents to start a `null` Entity.

Reply:
It's probably insignificant, really; feel free to ignore. I just wanted to confirm whether `flattener.getCurrentEntityName() != null` implies `path.contains(flattener.getEntityMarker())`, so that it _can't_ throw an AIOOBE anymore. I assume it does.

(Also, there's the question of nested entities (more than one entity marker in the path), but I won't get into it here...)